src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs

```diff
@@ -35,59 +73,130 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
 
         public virtual void EnqueueMessage(LogMessageEntry message)
         {
-            if (!_messageQueue.IsAddingCompleted)
-            {
-                try
-                {
-                    _messageQueue.Add(message);
-                    return;
-                }
-                catch (InvalidOperationException) { }
-            }
-
-            // Adding is completed so just log the message
-            try
+            // cannot enqueue when adding is completed
+            if (!Enqueue(message))
             {
                 WriteMessage(message);
             }
-            catch (Exception) { }
         }
 
         // for testing
         internal void WriteMessage(LogMessageEntry entry)
         {
-            IConsole console = entry.LogAsError ? ErrorConsole : Console;
-            console.Write(entry.Message);
+            try
+            {
+                IConsole console = entry.LogAsError ? ErrorConsole : Console;
+                console.Write(entry.Message);
+            }
+            catch (Exception)
+            {
+                CompleteAdding();
+            }
         }
 
         private void ProcessLogQueue()
         {
-            try
+            while (!_isAddingCompleted || _messageQueue.Count > 0)
             {
-                foreach (LogMessageEntry message in _messageQueue.GetConsumingEnumerable())
+                if (TryDequeue(out LogMessageEntry message))
                 {
                     WriteMessage(message);
                 }
             }
-            catch
+        }
+
+        public bool Enqueue(LogMessageEntry item)
+        {
+            lock (_messageQueue)
+            {
+                while (_messageQueue.Count >= MaxQueueLength && !_isAddingCompleted)
+                {
+                    if (FullMode == ConsoleLoggerBufferFullMode.DropWrite)
+                    {
+                        _messagesDropped++;
+                        return true;
+                    }
+
+                    Debug.Assert(FullMode == ConsoleLoggerBufferFullMode.Wait);
+                    Monitor.Wait(_messageQueue);
+                }
+
+                if (_messageQueue.Count < MaxQueueLength && !_isAddingCompleted)
+                {
```

Comment:
Can this just be:
```suggestion
                if (!_isAddingCompleted)
                {
                    Debug.Assert(_messageQueue.Count < MaxQueueLength);
```
?